File: ashcore/benchmarks/bench_sync.c
/*
 * Benchmark: pthread_mutex contention — 8 threads, 100k lock/unlock pairs each.
 * Equivalent to bench_sync.mojo.
 */
#include <stdio.h>
#include <pthread.h>
#include <time.h>

static long ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000000000L + ts.tv_nsec;
}

#define WORKERS 8
#define ITERS   100000

static pthread_mutex_t mu = PTHREAD_MUTEX_INITIALIZER;
static long long counter  = 0;

static void *worker(void *arg) {
    (void)arg;
    for (int i = 0; i < ITERS; i++) {
        pthread_mutex_lock(&mu);
        counter++;
        pthread_mutex_unlock(&mu);
    }
    return NULL;
}

int main(void) {
    pthread_t threads[WORKERS];

    long t0 = ns();
    for (int i = 0; i < WORKERS; i++) pthread_create(&threads[i], NULL, worker, NULL);
    for (int i = 0; i < WORKERS; i++) pthread_join(threads[i], NULL);
    long t1 = ns();

    long total_ops = (long)WORKERS * ITERS;
    printf("lock_ns=%ld\n",   t1 - t0);
    printf("total_ops=%ld\n", total_ops);
    printf("per_op_ns=%ld\n", (t1 - t0) / total_ops);
    printf("counter=%lld\n",  counter);
    printf("workers=%d\n",    WORKERS);
    return 0;
}

File: ashcore/benchmarks/bench_sync.py
"""
Benchmark: Python threading.Lock contention — 8 threads, 100k lock/unlock pairs.
Equivalent to bench_sync.mojo / bench_sync.c.
"""
import threading
import time

WORKERS = 8
ITERS   = 100_000

mu      = threading.Lock()
counter = 0

def worker():
    global counter
    for _ in range(ITERS):
        with mu:
            counter += 1

threads = [threading.Thread(target=worker) for _ in range(WORKERS)]

t0 = time.perf_counter_ns()
for t in threads: t.start()
for t in threads: t.join()
t1 = time.perf_counter_ns()

total_ops = WORKERS * ITERS
print(f"lock_ns={t1 - t0}")
print(f"total_ops={total_ops}")
print(f"per_op_ns={(t1 - t0) // total_ops}")
print(f"counter={counter}")
print(f"workers={WORKERS}")
